support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -114,61 +121,72 @@ class CreateZuoraSubscription(servicesProvider: ServiceProvider = ServiceProvide
-        soldToContact = state.user.deliveryAddress map (buildContactDetails(state.user, state.giftRecipient, _, state.user.deliveryInstructions)),
-        paymentMethod = state.paymentMethod.left.toOption,
-        subscriptionData = subscriptionData,
-        subscribeOptions = SubscribeOptions(billingEnabled, billingEnabled)
-      )
-    }
+    //Documentation for this request is here: https://www.zuora.com/developer/api-reference/#operation/Action_POSTsubscribe
+    SubscribeItem(
+      account = buildAccount(state),
+      billToContact = buildContactDetails(state.user, None, state.user.billingAddress),
+      soldToContact = state.user.deliveryAddress map (buildContactDetails(state.user, state.giftRecipient, _, state.user.deliveryInstructions)),
+      paymentMethod = state.paymentMethod.left.toOption,
+      subscriptionData = subscriptionData,
+      subscribeOptions = SubscribeOptions(generateInvoice = billingEnabled, processPayments = billingEnabled)
+    )
   }
 
-  private def buildSubscriptionData(state: CreateZuoraSubscriptionState, promotionService: PromotionService) = {
+  private def buildSubscriptionData(
+    state: CreateZuoraSubscriptionState,
+    promotionService: => PromotionService,
+    getCodeStatus: => GetCodeStatus,
+    now: () => LocalDate
+  ): Future[SubscriptionData] = {
     val isTestUser = state.user.isTestUser
     val config = zuoraConfigProvider.get(isTestUser)
-    val readerType: ReaderType = state.giftRecipient match  {
-      case _: Some[GiftRecipient] => ReaderType.Gift
-      case _ => ReaderType.Direct
-    }
     val stage = Configuration.stage
 
-    state.product match {
-      case c: Contribution => Right(c.build(state.requestId, config))
-      case d: DigitalPack => d.build(
-        state.requestId,
-        config,
-        state.user.billingAddress.country,
-        state.promoCode,
-        state.paymentMethod,
-        promotionService,
-        stage,
-        isTestUser
-      )
-      case p: Paper => p.build(
+    val eventualErrorOrSubscriptionData = state.product match {
+      case c: Contribution => EitherT.pure[Future, String](buildContributionSubscription(c, state.requestId, config))
+      case d: DigitalPack => buildDigitalPackSubscription(
+        d,
         state.requestId,
         state.user.billingAddress.country,
-        state.promoCode,
-        state.firstDeliveryDate,
+        state.paymentMethod.leftMap(_ => (config.digitalPack, state.promoCode)),
         promotionService,
         stage,
-        isTestUser
+        isTestUser,
+        getCodeStatus,
+        now
       )
-      case w: GuardianWeekly => w.build(
+      case p: Paper => EitherT.fromEither[Future](buildPaperSubscription(
+        p,
         state.requestId,
         state.user.billingAddress.country,
         state.promoCode,
         state.firstDeliveryDate,
         promotionService,
-        readerType,
         stage,
         isTestUser
-      )
+      ).leftMap(_.msg))
```

Comment:
we used to use `toString` but now we do `msg`.  The toString would have something like "NotApplicable" wheras the message is something like "This promotion is not applicable"  I hope this won't break something and hopefully it would be more friendly.